src/axom/inlet/tests/inlet_object.cpp

```diff
@@ -428,6 +426,114 @@ TEST(inlet_object, implicit_conversion_primitives)
   EXPECT_EQ(arr, expected_arr);
 }
 
+TEST(inlet_dict, basic_dicts)
+{
+  std::string testString =
+    "foo = { [\"key1\"] = 4, [\"key3\"] = 6, [\"key2\"] = 10}";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  inlet.addIntDict("foo", "foo's description");
+  std::unordered_map<std::string, int> dict = inlet["foo"];
+  std::unordered_map<std::string, int> correct = {{"key1", 4},
+                                                  {"key3", 6},
+                                                  {"key2", 10}};
+  EXPECT_EQ(dict, correct);
+}
+
+TEST(inlet_dict, simple_dict_of_struct_by_value)
+{
+  std::string testString =
+    "foo = { [\"key1\"] = { bar = true; baz = false}, "
+    "        [\"key2\"] = { bar = false; baz = true} }";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  auto& dict_table = inlet.addGenericDict("foo");
+
+  dict_table.addBool("bar", "bar's description");
+  dict_table.addBool("baz", "baz's description");
+  std::unordered_map<std::string, Foo> expected_foos = {{"key1", {true, false}},
+                                                        {"key2", {false, true}}};
+  std::unordered_map<std::string, Foo> foos;
+  foos = inlet["foo"].get<std::unordered_map<std::string, Foo>>();
```

Comment:
Just to clarify, with this new PR, there are separate `get<>` calls to access the integer-indexed entries and the string-indexed entries, right?  (w/ an `unordered_map<int, T>` and an `unordered_map<std::string, T>` template type)

Reply:
Follow-up 1: Should there be a test with a dictionary that has both int-valued and string-valued indices?

Reply:
Follow-up 2: Does the struct type for the string-indexed entries have to match the struct for the int-indexed entries?

Reply:
Would you want to support both index types in a single container? I know it's supported in Lua but I think I find it very confusing from user stand-point.

For example:

```
foo = { ["1"] = "something", [1] = "something else"}
```

Reply:
Right now this is not supported - I think it would require something like an `unordered_map<variant<int, std::string>, T>` which is definitely possible and relatively easy to use:
```cpp
    std::unordered_map<std::variant<std::string, int>, std::string> map;
    map[0] = "int-valued";
    map["3"] = "string-valued";
    for (const auto& thing : map)
    {
        std::visit([](const auto& key) {std::cout << key;}, thing.first);
        std::cout << ": " << thing.second << "\n";
    }
// output:
3: string-valued
0: int-valued
```
This would take a little bit of work to implement in C++11/14 so it may belong in a separate PR.

Reply:
@white238  re: dictionaries with `"1"` and `1` as keys -- I agree that `inlet` should only support a subset of Lua and that the particular use case that I pointed out (string versions of ints) would be an antipattern that we would not want to support (or at least promote).

My concern was more in the realm of error checking / handling -- i.e. what would happen if a user attempted to do that.  Would the code crash? Would it silently overwrite one of the entries (and which one)? etc...

Reply:
Given `foo = { ["1"] = "something", [1] = "something else"}`, the answer depends on whether `foo` was added as a dictionary or as an array.

If it was added as a dictionary, the integer-valued entry is ignored completely.  Inlet doesn't warn or error in this case - my understanding is that this was done so other things could be present in a table containing keyed entries:
```lua
foo  = {
    [1] = "something",
    bar = true,
}
```

If it was added as an array, an exception will be thrown by `std::stoi` when the key is converted to an integer.  There was some opposition to exceptions earlier so I've replaced `std::stoi` with the (somewhat returncode-based) libc `strtol` and added a SLIC error message on failure.